### ISSUE TITLE
blkid reports whole disk is zfs pool 

### ISSUE THREAD

**Issue by olivier-klein** [zfsonlinux/zfs#9069]
Issue raised in
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=888114#25

I am experiencing the same bug (ubuntu 18.04 / manjaro 18) fdisk shows correctly the partition table but not gparted (or it seems blkid, in my case util-linux 2.34 14-Jun-2019). The whole disk /dev/nvme0n1 is TYPE='zfs member'.If that can be useful I noticed this bug was not present went the zfs partition was freshly created by following the instructions on https://wiki.archlinux.org/index.php/ZFS.A launch of gparted after the zpool create showed properly the partition table, even flagging as zfs type the new poolIt appears only after enabling the automatic zfs mounting on boot, and rebooting the machine.
--

**Comment by nivedita76** [comment]
FWIW, this doesn't happen for me. I don't have gparted but blkid -p does not show whole disk as zfs. I have util-linux 2.33.2 on a gentoo system. I'm not sure what Ubuntu's auto mount does, I am mounting using a custom init script in a custom initramfs, using zpool import -N -R followed by zfs mount. So I'm not using a cache file but surely that shouldn't affect blkid.

`$ zpool status
  pool: rani
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
        still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
        the pool may no longer be accessible by software that does not support
        the features. See zpool-features(5) for details.
  scan: scrub repaired 0B in 0 days 00:22:26 with 0 errors on Fri Jul 26 18:35:21 2019
config:

        NAME           STATE     READ WRITE CKSUM
        rani           ONLINE       0     0     0
          mirror-0     ONLINE       0     0     0
            nvme0n1p4  ONLINE       0     0     0
            nvme1n1p4  ONLINE       0     0     0
          mirror-1     ONLINE       0     0     0
            nvme2n1p4  ONLINE       0     0     0
            nvme3n1p4  ONLINE       0     0     0

errors: No known data errors`

$ sudo blkid -p /dev/nvme0n1
/dev/nvme0n1: PTUUID="1bc79853-ad42-4078-8d2a-774501c3d895" PTTYPE="gpt"

**Comment by olivier-klein** [comment]
It seems to me that this bug has been reported elsewhere:

https://gitlab.gnome.org/GNOME/gparted/issues/14
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=888114
https://bbs.archlinux.org/viewtopic.php?id=206121
https://bbs.archlinux.org/viewtopic.php?id=202587

Maybe it is a bug linked to zfs. I have opened a bug report on zfsonlinux: https://github.com/zfsonlinux/zfs/issues/9105

Is it possible that it affects only SSD?

The question is how to solve it? It seems that wipefs is useless. I have not tried it.

**Comment by nivedita76** [comment]
> It seems to me that this bug has been reported elsewhere:
> 
> https://gitlab.gnome.org/GNOME/gparted/issues/14
> https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=888114
> https://bbs.archlinux.org/viewtopic.php?id=206121
> https://bbs.archlinux.org/viewtopic.php?id=202587
> 
> Maybe it is a bug linked to zfs. I have opened a bug report on zfsonlinux: #9105
> 
> Is it possible that it affects only SSD?
> 
> The question is how to solve it? It seems that wipefs is useless. I have not tried it.

Why did you open another bug report? What’s wrong with this one?
I know it has been reported before what I’m saying as a data point is that it doesn’t happen in my system, which is running zfs git on NVMe ssds.

**Comment by olivier-klein** [comment]
Sorry for the confusion. I was just trying to be more articulate about it.
Let me more explicit about the SSD: I am wondering if the culprit is the fact that the zpool was created with ashift=13?

The exact command that I used to create the pool is the following:
# sudo zpool create -f -o ashift=13 -m /mnt/tank tank nvmePC401_NVMe_SK_hynix_1TB_MI93T003810403E62-part5

**Comment by olivier-klein** [comment]
For information; zpool is fine... It is the other patitions that are messed up.
#zpool status
  pool: tank
 state: ONLINE
  scan: none requested
config:

	NAME                                                    STATE     READ WRITE CKSUM
	tank                                                    ONLINE       0     0     0
	  nvme-PC401_NVMe_SK_hynix_1TB_MI93T003810403E62-part5  ONLINE       0     0     0

errors: No known data errors

**Comment by nivedita76** [comment]
Is this actually pasted correctly from your new issue? p5's nvme namespace doesn't match. Did you ever have a zfs pool on this SSD before? From the previous bug reports it sounds like things get confused because not all the zfs signatures get removed by wipefs etc, though not sure that's still an issue. I think mine was a clean install on previously unused disks though I can't exactly remember now given how long ago it was.

#lsblk -a
nvme0**m**1 259:0 0 953,9G 0 disk
├─nvme0m1p1 259:1 0 780M 0 part /boot/efi
├─nvme0m1p2 259:2 0 5G 0 part
├─nvme0m1p3 259:3 0 97,7G 0 part
├─nvme0m1p4 259:4 0 97,7G 0 part /
└─nvme0**n**1p5 259:5 0 752,8G 0 part

**Comment by olivier-klein** [comment]
Update about the issue. I tried yesterday to reinstall completely the OS. The hope was to wipe out compeltely the zfs_member signature from the /dev/nvme01 SSD. I tried different strategy but did not go as far as to zero the HD with dd. Thus I got stuck with the FSTYPE zfs_member put on the whole disk.  Never the less the boot recognizes properly the partition table (fdisk and parted are not bothered by this problem). Thus I came back to square 1.

The update configuration indicates the following
lsblk -f
NAME        FSTYPE     LABEL   UUID                                 FSAVAIL FSUSE% MOUNTPOINT
nvme0n1     zfs_member tank    9593340550191022900                                 
├─nvme0n1p1 vfat       ESP     F8E8-2858                             738,1M     5% /boot/efi
├─nvme0n1p2 vfat       OS      5224-B9FA                                           
├─nvme0n1p3 ext4       UBUNTU  bed2f836-754b-477b-8bdb-3cba7d56fae3                
├─nvme0n1p4 ext4       MANJARO 1b4a8f02-878e-4870-a54c-3a08f7b5688a     76G    15% /
└─nvme0n1p5 zfs_member tank    9593340550191022900 

FYI there was no zfs install prior to the emergence of the problem (and indeed I mispelled in the previous paste nvme0n1 by nvme0m1).
Interestingly, and this occured before, nvme0n1p4 was also labelled tank (same as disk and partition 5)  after the fresh install (this also occured initially). It did not appear in the signature because I changed the label manually with e2label.

Back to a stage where I could reimport the zfs pool and no boot message error. (I suspect now that my boot problem came from a different conflict between powertop and the messed up disk partition).

In principle, the problem should be reproducible. I have kept a track of all what I did initally and the logs  (the procedure follows the https://wiki.archlinux.org/index.php/ZFS instruction).

1) create the zpool with by_id :
#zpool create -f -o ashift=13 -m /mnt/tank tank nvmePC401_NVMe_SK_hynix_1TB_MI93T003810403E62-part5

-> Note that I used here ashift=13 for SSD
-> at this stage gparted indicated a proper partition table

2) automatic start
#zpool set cachefile=/etc/zfs/zpool.cache tank   
-> I suspect that the problem was triggered there upon reboot

3) enable the service
systemctl enable zfs-import-cache
systemctl enable zfs-import.target
systemctl enable zfs-mount
systemctl enable zfs.target

**Comment by s-m-e** [comment]
I am still seeing this bug in `blkid` (through gparted). There was an [email sent to the util-linux mailing list](https://marc.info/?l=util-linux-ng&m=156491424909123&w=2), without a reply though. Is anyone looking any further into this?

**Comment by satmandu** [comment]
This is also mentioned here: https://bugs.launchpad.net/ubuntu/+source/util-linux/+bug/1808421

**Comment by deliciouslytyped** [comment]
I just ran into this issue. I may or may not follow up with adding additional information to this post.

**Comment by s-m-e** [comment]
I soon need to upgrade a bunch of systems where ZFS "shares" SSD drives with other partitions. I am affected by this bug on all of them. My best bad idea is to wipe the drives and start from scratch as must Linux installers rely on `blkid` to work in one way or the other. Any better idea how I can preserve my zpools in the process? Could I somehow manually and temporarily remove the `FSTYPE zfs_member` state from my disks without damaging stuff?

**Comment by s-m-e** [comment]
Not stale by any means, still present.

**Comment by mfleetwo** [comment]
Upstream root cause and fix:
* [util-linux issue 918 - blkid reports disk as zfs_member if it has a zfs_member partition](https://github.com/util-linux/util-linux/issues/918)
* Fixed by commit [libblkid: (zfs) don't probe whole-disk areas covered by partitions](https://github.com/util-linux/util-linux/commit/f6e182078a3f05af3ce3b5e5a2d1956d4b9677e2)
* First included in [util-linux v2.35](https://github.com/util-linux/util-linux/blob/master/Documentation/releases/v2.35-ReleaseNotes#L231) released [Jan 21 2020](https://github.com/util-linux/util-linux/releases/tag/v2.35)
